### DISCUSSION TITLE
Send emails with an SMTP server

### DISCUSSION THREAD

**Discussion by CogalTek** [pocketbase/pocketbase#3817] [Q&A] [answered]
Hello, I hope you are well;

today I tackled sending emails, so I intend to create a route in the API which would allow me to send emails,

So I created in pb_hooks: main.pb.js

```Javascript
const nodemailer = require('nodemailer');

let transporter = nodemailer.createTransport({
    host: 'adresse',
    port: 465,
    secure: true,
    auth: {
        user: 'mercure@gestionbat.fr',
        pass: 'password'
    }
});

let mailOptions = {
    from: '"Mercure" <mercure@gestionbat.fr>',
    to: 'mathieu.rio@outlook.com',
    subject: 'Hello',
    text: 'Hello world?',
    html: '<b>Hello world?</b>'
};

routerAdd("GET", "/hello", (c) => {
    transporter.sendMail(mailOptions, (error, info) => {
        if (error) {
            return console.log(error);
        }
        console.log('Message sent: %s', info.messageId);
    });
    return c.string(200, "Hello, World!");
}, requireRecordAuth());
```

the problem is that when I tried to import "nodemailer" it returned an error:

```
File pb_hooks/main.pb.js changed, restarting...
Failed to execute main.pb.js:
 - GoError: Invalid module at github.com/dop251/goja_nodejs/require.(*RequireModule).require-fm (native)
2023/11/23 14:25:37 Server started at http://127.0.0.1:8090
├─ REST API: http://127.0.0.1:8090/api/
└─ Admin UI: http://127.0.0.1:8090/_/
```

so I did an npm install nodemailer and that didn't fix anything,

I admit I don't understand why I have this error, would it be possible to get help?

oh and if I manage to resolve this problem I would like to be able to send emails with the pocketbase connection password (it's the same as for the SMTP server for each account) but I wouldn't be able to use it if it is entered as a password in pocketbase because it will be encrypted, but I don't want to leave it written in plain text either, do you have any ideas?

**Comment by ganigeorgiev** [comment]
External modules can be loaded only if they don't rely on runtime specific apis, aka. they are pure ES5 library (see this and other limitations in https://pocketbase.io/docs/js-overview/#caveats-and-limitations).

With that said, for sending emails with the JS `pb_hooks` you can use the builtin PocketBase mail client - https://pocketbase.io/docs/js-sending-emails/.

**Reply by CogalTek** [reply]
ok thank you for your answer, the problem is that each user of the site must be able to send emails with their own email address, does pocketbase do this?

**Reply by ganigeorgiev** [reply]
@CogalTek I'm not sure that I understand what do you mean.

Please see the [second link](https://pocketbase.io/docs/js-sending-emails/#send-custom-email). You can change the `from.address` field to anything you want.

Whether that would be valid it is up to your mail configuration.
If you don't have control over the user email domain, aka. every random user can register, then you can't send email pretending to be the user (_technically the email will be submitted but most likely will be immediately flagged as it will fail the sender DMARC/DKIM/SPF/etc. DNS records checks_). Or in other words, the `from` address must be something you are in control.

**Reply by CogalTek** [reply]
ok, I understand,

actually I have an SMTP server

email addresses and passwords have been created for this SMTP server,

and so to send an email I have to connect to the server with a password and an email address.

the problem I have with the pocketbase email system is that it does not offer me the option to send an email with a password and an email that differs depending on all the users.

I don't know if this is understandable

**Reply by ganigeorgiev** [reply, marked as answer]
At the moment the JS `pb_hooks` doesn't expose methods to create smtp clients with different settings then what is stored in `$app.settings()`.

If you use Go, you can initialize [`mailer.SmtpClient`](https://pkg.go.dev/github.com/pocketbase/pocketbase@v0.19.4/tools/mailer#SmtpClient) for each of your different configurations:

```go
client := mailer.SmtpClient{
    Host:       "",
    Port:       587,
    Username:   "",
    Password:   "",
}
```

I guess it can be added to the JS `pb_hooks` but I don't fully understand the use case. Could you elaborate why do you want to do this?

**Reply by CogalTek** [reply]
I am creating a tool that should allow all employees of a company to connect and send emails to many people and therefore when an email is sent it must be in the name of the employee

**Reply by ganigeorgiev** [reply]
@CogalTek Sorry, I'm failing to understand why do you want to do this. How do you even store the different passwords, as plain text (aka. is it ok admins to see the passwords)?

If you are in control of the SMTP server and the users domain(s), then it doesn't matter with what credentials the mail client authenticated against the SMTP server and you can make any `*@yourdomain.com` from addresses to be considered valid (you can change the `from.address` when sending the mail).

**Comment by Lavis166** [comment]
Currently, the **JS pb_hooks** doesn't allow you to create SMTP clients with different settings from what’s saved in `$app.settings()`.

However, if you’re using Go, you can set up a custom **SMTP client** like this:

```go
client := mailer.SmtpClient{
    Host:       "",
    Port:       587,
    Username:   "",
    Password:   "",
}
```

It’s possible to add this feature to the **JS pb_hooks**, but I’m not entirely sure about the reason for it. Could you explain your needs a bit more?

For SMTP server options, you could consider services like **SMTPmart**, **Mailgun** or **SMTPget**.

**Reply by ganigeorgiev** [reply]
Please don't spam unnecessary old discussion.
